File: java/ai/src/main/java/com/google/genkit/ai/evaluation/Evaluator.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.google.genkit.ai.evaluation;

import java.util.*;
import java.util.function.Consumer;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.genkit.core.*;

/**
 * Evaluator represents an evaluation action that assesses the quality of AI
 * outputs.
 * 
 * <p>
 * Evaluators are a core primitive in Genkit that allow you to measure and track
 * the quality of your AI applications. They can be used to:
 * <ul>
 * <li>Score outputs based on various criteria (accuracy, relevance, etc.)</li>
 * <li>Compare outputs against reference data</li>
 * <li>Run automated quality checks in CI/CD pipelines</li>
 * <li>Monitor production quality over time</li>
 * </ul>
 *
 * @param <O>
 *            the type of evaluator-specific options
 */
public class Evaluator<O> implements Action<EvalRequest, List<EvalResponse>, Void> {

  private static final Logger logger = LoggerFactory.getLogger(Evaluator.class);

  public static final String METADATA_KEY_DISPLAY_NAME = "evaluatorDisplayName";
  public static final String METADATA_KEY_DEFINITION = "evaluatorDefinition";
  public static final String METADATA_KEY_IS_BILLED = "evaluatorIsBilled";

  private final String name;
  private final EvaluatorInfo info;
  private final EvaluatorFn<O> evaluatorFn;
  private final Class<O> optionsClass;
  private final ActionDesc desc;

  private Evaluator(Builder<O> builder) {
    this.name = builder.name;
    this.info = EvaluatorInfo.builder().displayName(builder.displayName).definition(builder.definition)
        .isBilled(builder.isBilled).build();
    this.evaluatorFn = builder.evaluatorFn;
    this.optionsClass = builder.optionsClass;

    // Build metadata
    Map<String, Object> metadata = new HashMap<>();
    metadata.put("type", "evaluator");
    Map<String, Object> evaluatorMetadata = new HashMap<>();
    evaluatorMetadata.put(METADATA_KEY_DISPLAY_NAME, builder.displayName);
    evaluatorMetadata.put(METADATA_KEY_DEFINITION, builder.definition);
    evaluatorMetadata.put(METADATA_KEY_IS_BILLED, builder.isBilled);
    metadata.put("evaluator", evaluatorMetadata);

    this.desc = ActionDesc.builder().type(ActionType.EVALUATOR).name(name).description(builder.definition)
        .metadata(metadata).build();
  }

  /**
   * Creates a new Evaluator builder.
   *
   * @param <O>
   *            the options type
   * @return a new builder
   */
  public static <O> Builder<O> builder() {
    return new Builder<>();
  }

  /**
   * Defines a new evaluator and registers it with the registry.
   *
   * @param <O>
   *            the options type
   * @param registry
   *            the registry to register with
   * @param name
   *            the evaluator name
   * @param displayName
   *            the display name shown in the UI
   * @param definition
   *            description of what the evaluator measures
   * @param evaluatorFn
   *            the evaluation function
   * @return the created evaluator
   */
  public static <O> Evaluator<O> define(Registry registry, String name, String displayName, String definition,
      EvaluatorFn<O> evaluatorFn) {
    return define(registry, name, displayName, definition, true, null, evaluatorFn);
  }

  /**
   * Defines a new evaluator with full options and registers it with the registry.
   *
   * @param <O>
   *            the options type
   * @param registry
   *            the registry to register with
   * @param name
   *            the evaluator name
   * @param displayName
   *            the display name shown in the UI
   * @param definition
   *            description of what the evaluator measures
   * @param isBilled
   *            whether using this evaluator incurs costs
   * @param optionsClass
   *            the class for evaluator-specific options
   * @param evaluatorFn
   *            the evaluation function
   * @return the created evaluator
   */
  public static <O> Evaluator<O> define(Registry registry, String name, String displayName, String definition,
      boolean isBilled, Class<O> optionsClass, EvaluatorFn<O> evaluatorFn) {

    Evaluator<O> evaluator = Evaluator.<O>builder().name(name).displayName(displayName).definition(definition)
        .isBilled(isBilled).optionsClass(optionsClass).evaluatorFn(evaluatorFn).build();

    evaluator.register(registry);
    return evaluator;
  }

  @Override
  public String getName() {
    return name;
  }

  @Override
  public ActionType getType() {
    return ActionType.EVALUATOR;
  }

  @Override
  public ActionDesc getDesc() {
    return desc;
  }

  /**
   * Gets the evaluator info.
   *
   * @return the evaluator info
   */
  public EvaluatorInfo getInfo() {
    return info;
  }

  @Override
  public List<EvalResponse> run(ActionContext ctx, EvalRequest input) throws GenkitException {
    return run(ctx, input, null);
  }

  @Override
  public List<EvalResponse> run(ActionContext ctx, EvalRequest input, Consumer<Void> streamCallback)
      throws GenkitException {
    List<EvalResponse> responses = new ArrayList<>();
    List<EvalDataPoint> dataset = input.getDataset();

    if (dataset == null || dataset.isEmpty()) {
      return responses;
    }

    int batchSize = input.getBatchSize() != null ? input.getBatchSize() : 10;

    // Process in batches
    List<List<EvalDataPoint>> batches = batchList(dataset, batchSize);
    int sampleIndex = 0;

    for (List<EvalDataPoint> batch : batches) {
      for (EvalDataPoint dataPoint : batch) {
        try {
          @SuppressWarnings("unchecked")
          O options = input.getOptions() != null && optionsClass != null
              ? JsonUtils.getObjectMapper().convertValue(input.getOptions(), optionsClass)
              : null;

          EvalResponse response = evaluatorFn.evaluate(dataPoint, options);
          if (response.getSampleIndex() == null) {
            response.setSampleIndex(sampleIndex);
          }
          responses.add(response);
        } catch (Exception e) {
          logger.error("Error evaluating data point: {}", dataPoint.getTestCaseId(), e);
          // Create an error response
          Score errorScore = Score.builder().error(e.getMessage()).status(EvalStatus.UNKNOWN).build();
          responses.add(EvalResponse.builder().testCaseId(dataPoint.getTestCaseId()).sampleIndex(sampleIndex)
              .evaluation(errorScore).build());
        }
        sampleIndex++;
      }
    }

    return responses;
  }

  @Override
  public JsonNode runJson(ActionContext ctx, JsonNode input, Consumer<JsonNode> streamCallback)
      throws GenkitException {
    EvalRequest request = JsonUtils.fromJsonNode(input, EvalRequest.class);
    List<EvalResponse> result = run(ctx, request, null);
    return JsonUtils.toJsonNode(result);
  }

  @Override
  public ActionRunResult<JsonNode> runJsonWithTelemetry(ActionContext ctx, JsonNode input,
      Consumer<JsonNode> streamCallback) throws GenkitException {
    JsonNode result = runJson(ctx, input, streamCallback);
    return new ActionRunResult<>(result, null, null);
  }

  @Override
  public void register(Registry registry) {
    String key = ActionType.EVALUATOR.keyFromName(name);
    registry.registerAction(key, this);
    logger.info("Registered evaluator: {}", key);
  }

  @Override
  public Map<String, Object> getInputSchema() {
    // Define the input schema for EvalRequest
    Map<String, Object> schema = new HashMap<>();
    schema.put("type", "object");

    Map<String, Object> properties = new HashMap<>();

    // dataset property
    Map<String, Object> datasetProp = new HashMap<>();
    datasetProp.put("type", "array");
    Map<String, Object> dataPointSchema = new HashMap<>();
    dataPointSchema.put("type", "object");
    datasetProp.put("items", dataPointSchema);
    properties.put("dataset", datasetProp);

    // evalRunId property
    Map<String, Object> evalRunIdProp = new HashMap<>();
    evalRunIdProp.put("type", "string");
    properties.put("evalRunId", evalRunIdProp);

    // batchSize property
    Map<String, Object> batchSizeProp = new HashMap<>();
    batchSizeProp.put("type", "integer");
    properties.put("batchSize", batchSizeProp);

    schema.put("properties", properties);
    schema.put("required", Arrays.asList("dataset"));

    return schema;
  }

  @Override
  public Map<String, Object> getOutputSchema() {
    // Define the output schema for List<EvalResponse>
    Map<String, Object> schema = new HashMap<>();
    schema.put("type", "array");

    Map<String, Object> itemSchema = new HashMap<>();
    itemSchema.put("type", "object");

    Map<String, Object> itemProps = new HashMap<>();

    // testCaseId
    Map<String, Object> testCaseIdProp = new HashMap<>();
    testCaseIdProp.put("type", "string");
    itemProps.put("testCaseId", testCaseIdProp);

    // evaluation (Score)
    Map<String, Object> evaluationProp = new HashMap<>();
    evaluationProp.put("type", "object");
    itemProps.put("evaluation", evaluationProp);

    itemSchema.put("properties", itemProps);
    schema.put("items", itemSchema);

    return schema;
  }

  @Override
  public Map<String, Object> getMetadata() {
    return desc.getMetadata();
  }

  /**
   * Splits a list into batches.
   */
  private static <T> List<List<T>> batchList(List<T> list, int batchSize) {
    List<List<T>> batches = new ArrayList<>();
    for (int i = 0; i < list.size(); i += batchSize) {
      batches.add(list.subList(i, Math.min(i + batchSize, list.size())));
    }
    return batches;
  }

  /**
   * Builder for creating Evaluator instances.
   *
   * @param <O>
   *            the options type
   */
  public static class Builder<O> {
    private String name;
    private String displayName;
    private String definition;
    private boolean isBilled = true;
    private Class<O> optionsClass;
    private EvaluatorFn<O> evaluatorFn;

    public Builder<O> name(String name) {
      this.name = name;
      return this;
    }

    public Builder<O> displayName(String displayName) {
      this.displayName = displayName;
      return this;
    }

    public Builder<O> definition(String definition) {
      this.definition = definition;
      return this;
    }

    public Builder<O> isBilled(boolean isBilled) {
      this.isBilled = isBilled;
      return this;
    }

    public Builder<O> optionsClass(Class<O> optionsClass) {
      this.optionsClass = optionsClass;
      return this;
    }

    public Builder<O> evaluatorFn(EvaluatorFn<O> evaluatorFn) {
      this.evaluatorFn = evaluatorFn;
      return this;
    }

    public Evaluator<O> build() {
      if (name == null || name.isEmpty()) {
        throw new IllegalArgumentException("Evaluator name is required");
      }
      if (displayName == null) {
        displayName = name;
      }
      if (definition == null) {
        definition = "";
      }
      if (evaluatorFn == null) {
        throw new IllegalArgumentException("Evaluator function is required");
      }
      return new Evaluator<>(this);
    }
  }
}

File: py/packages/genkit/src/genkit/ai/evaluator.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Evaluator type definitions for the Genkit framework.

This module defines the type interfaces for evaluators in the Genkit framework.
Evaluators are used for assessing the quality of output of a Genkit flow or
model.
"""

import traceback
import uuid
from collections.abc import Callable, Coroutine
from typing import Any, ClassVar, TypeVar, cast

from pydantic import BaseModel, ConfigDict
from pydantic.alias_generators import to_camel

from genkit.core.action import Action, ActionMetadata
from genkit.core.action import ActionKind
from genkit.core._internal._logging import get_logger
from genkit.core._internal._registry import Registry
from genkit.core._internal._schema import to_json_schema
from genkit.core._internal._typing import (
    BaseDataPoint,
    EvalFnResponse,
    EvalRequest,
    EvalResponse,
    EvalStatusEnum,
    Score,
    SpanMetadata,
)
from genkit.core.tracing import run_in_new_span

logger = get_logger(__name__)

EVALUATOR_METADATA_KEY_DISPLAY_NAME = 'evaluatorDisplayName'
EVALUATOR_METADATA_KEY_DEFINITION = 'evaluatorDefinition'
EVALUATOR_METADATA_KEY_IS_BILLED = 'evaluatorIsBilled'

T = TypeVar('T')

# User-provided evaluator function that evaluates a single datapoint.
# Must be async (coroutine function).
EvaluatorFn = Callable[[BaseDataPoint, T], Coroutine[Any, Any, EvalFnResponse]]

# User-provided batch evaluator function that evaluates an EvaluationRequest
BatchEvaluatorFn = Callable[[EvalRequest, T], Coroutine[Any, Any, list[EvalFnResponse]]]


class EvaluatorRef(BaseModel):
    """Reference to an evaluator."""

    model_config: ClassVar[ConfigDict] = ConfigDict(extra='forbid', populate_by_name=True, alias_generator=to_camel)

    name: str
    config_schema: dict[str, object] | None = None


def evaluator_ref(name: str, config_schema: dict[str, object] | None = None) -> EvaluatorRef:
    """Create a reference to an evaluator.

    Args:
        name: Name of the evaluator.
        config_schema: Optional schema for evaluator configuration.

    Returns:
        An EvaluatorRef instance.
    """
    return EvaluatorRef(name=name, config_schema=config_schema)


def evaluator_action_metadata(
    name: str,
    config_schema: type | dict[str, Any] | None = None,
) -> ActionMetadata:
    """Generates an ActionMetadata for evaluators.

    Args:
        name: Name of the evaluator.
        config_schema: Optional schema for evaluator configuration.

    Returns:
        An ActionMetadata instance for the evaluator.
    """
    return ActionMetadata(
        kind=cast(ActionKind, ActionKind.EVALUATOR),
        name=name,
        input_json_schema=to_json_schema(EvalRequest),
        output_json_schema=to_json_schema(list[EvalFnResponse]),
        metadata={'evaluator': {'customOptions': to_json_schema(config_schema) if config_schema else None}},
    )


def _get_func_description(func: Callable[..., Any], description: str | None = None) -> str:
    """Get the description of a function.

    Args:
        func: The function to get the description of.
        description: The description to use if the function docstring is
            empty.
    """
    if description is not None:
        return description
    if func.__doc__ is not None:
        return func.__doc__
    return ''


def define_evaluator(
    registry: Registry,
    name: str,
    display_name: str,
    definition: str,
    fn: EvaluatorFn[Any],
    is_billed: bool = False,
    config_schema: type[BaseModel] | dict[str, object] | None = None,
    metadata: dict[str, object] | None = None,
    description: str | None = None,
) -> Action:
    """Define an evaluator action.

    This action runs the callback function on the every sample of
    the input dataset.

    Args:
        registry: The registry to register the evaluator in.
        name: Name of the evaluator.
        display_name: User-visible display name.
        definition: User-visible evaluator definition.
        fn: Function implementing the evaluator behavior.
        is_billed: Whether the evaluator performs any billed actions
            (paid APIs, LLMs etc.)
        config_schema: Optional schema for evaluator configuration.
        metadata: Optional metadata for the evaluator.
        description: Optional description for the evaluator.

    Returns:
        The registered evaluator Action.
    """
    evaluator_meta: dict[str, object] = dict(metadata) if metadata else {}
    evaluator_info: dict[str, object]
    existing_evaluator = evaluator_meta.get('evaluator')
    if isinstance(existing_evaluator, dict):
        evaluator_info = {str(key): value for key, value in existing_evaluator.items()}
    else:
        evaluator_info = {}
    evaluator_meta['evaluator'] = evaluator_info
    evaluator_info[EVALUATOR_METADATA_KEY_DEFINITION] = definition
    evaluator_info[EVALUATOR_METADATA_KEY_DISPLAY_NAME] = display_name
    evaluator_info[EVALUATOR_METADATA_KEY_IS_BILLED] = is_billed
    label_value = evaluator_info.get('label')
    if not isinstance(label_value, str) or not label_value:
        evaluator_info['label'] = name
    if config_schema:
        evaluator_info['customOptions'] = to_json_schema(config_schema)

    evaluator_description = _get_func_description(fn, description)

    async def eval_stepper_fn(req: EvalRequest) -> EvalResponse:
        eval_responses: list[EvalFnResponse] = []
        for index in range(len(req.dataset)):
            datapoint = req.dataset[index]
            if datapoint.test_case_id is None:
                datapoint.test_case_id = str(uuid.uuid4())
            span_metadata = SpanMetadata(
                name=f'Test Case {datapoint.test_case_id}',
                metadata={'evaluator:evalRunId': req.eval_run_id},
            )
            try:
                # Try to run with tracing, but fallback if tracing infrastructure fails
                # (e.g., in environments with NonRecordingSpans like pre-commit)
                try:
                    with run_in_new_span(span_metadata, labels={'genkit:type': 'evaluator'}) as span:
                        span_id = span.span_id
                        trace_id = span.trace_id
                        try:
                            span.set_input(datapoint)
                            test_case_output = await fn(datapoint, req.options)
                            test_case_output.span_id = span_id
                            test_case_output.trace_id = trace_id
                            span.set_output(test_case_output)
                            eval_responses.append(test_case_output)
                        except Exception as e:
                            logger.debug(f'eval_stepper_fn error: {e!s}')
                            logger.debug(traceback.format_exc())
                            evaluation = Score(
                                error=f'Evaluation of test case {datapoint.test_case_id} failed: \n{e!s}',
                                status=cast(EvalStatusEnum, EvalStatusEnum.FAIL),
                            )
                            eval_responses.append(
                                # The ty type checker only recognizes aliases, so we use them
                                # to pass both ty check and runtime validation.
                                EvalFnResponse(
                                    span_id=span_id,
                                    trace_id=trace_id,
                                    test_case_id=datapoint.test_case_id,
                                    evaluation=evaluation,
                                )
                            )
                            # Raise to mark span as failed
                            raise e
                except (AttributeError, UnboundLocalError):
                    # Fallback: run without span
                    try:
                        test_case_output = await fn(datapoint, req.options)
                        eval_responses.append(test_case_output)
                    except Exception as e:
                        logger.debug(f'eval_stepper_fn error: {e!s}')
                        logger.debug(traceback.format_exc())
                        evaluation = Score(
                            error=f'Evaluation of test case {datapoint.test_case_id} failed: \n{e!s}',
                            status=cast(EvalStatusEnum, EvalStatusEnum.FAIL),
                        )
                        eval_responses.append(
                            EvalFnResponse(
                                test_case_id=datapoint.test_case_id,
                                evaluation=evaluation,
                            )
                        )
            except Exception:  # noqa: S112 - intentionally continue processing other datapoints
                # Continue to process other points
                continue
        return EvalResponse(eval_responses)

    return registry.register_action(
        name=name,
        kind=cast(ActionKind, ActionKind.EVALUATOR),
        fn=eval_stepper_fn,
        metadata=evaluator_meta,
        description=evaluator_description,
    )


def define_batch_evaluator(
    registry: Registry,
    name: str,
    display_name: str,
    definition: str,
    fn: BatchEvaluatorFn[Any],
    is_billed: bool = False,
    config_schema: type[BaseModel] | dict[str, object] | None = None,
    metadata: dict[str, object] | None = None,
    description: str | None = None,
) -> Action:
    """Define a batch evaluator action.

    This action runs the callback function on the entire dataset.

    Args:
        registry: The registry to register the evaluator in.
        name: Name of the evaluator.
        display_name: User-visible display name.
        definition: User-visible evaluator definition.
        fn: Function implementing the evaluator behavior.
        is_billed: Whether the evaluator performs any billed actions
            (paid APIs, LLMs etc.)
        config_schema: Optional schema for evaluator configuration.
        metadata: Optional metadata for the evaluator.
        description: Optional description for the evaluator.

    Returns:
        The registered evaluator Action.
    """
    evaluator_meta: dict[str, object] = metadata.copy() if metadata else {}
    if 'evaluator' not in evaluator_meta:
        evaluator_meta['evaluator'] = {}
    # Cast to dict for nested operations - pyrefly doesn't narrow nested dict types
    evaluator_dict = cast(dict[str, object], evaluator_meta['evaluator'])
    evaluator_dict[EVALUATOR_METADATA_KEY_DEFINITION] = definition
    evaluator_dict[EVALUATOR_METADATA_KEY_DISPLAY_NAME] = display_name
    evaluator_dict[EVALUATOR_METADATA_KEY_IS_BILLED] = is_billed
    if 'label' not in evaluator_dict or not evaluator_dict['label']:
        evaluator_dict['label'] = name
    if config_schema:
        evaluator_dict['customOptions'] = to_json_schema(config_schema)

    evaluator_description = _get_func_description(fn, description)
    return registry.register_action(
        name=name,
        kind=cast(ActionKind, ActionKind.EVALUATOR),
        fn=fn,
        metadata=evaluator_meta,
        description=evaluator_description,
    )
